### PR TITLE
Traits for AM cloud

### PR DESCRIPTION
**Pull request by shrouti1507**
## Description of the change

> Description here

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: v0/destinations/am/transform.js [hunk 1]
```diff
@@ -516,17 +516,75 @@ function trackRevenueEvent(message, destination) {
...
+  let propertyToUpdate;
+  if (traitObject.hasOwnProperty(property)) {
+    propertyToUpdate = traitObject[property];
+    newProperty[property] = propertyToUpdate;
+    delete traitObject[property];
+  }
+  return traitObject;
+}
+
+function passingTraitConfigs(
```
Comment (prabrishac) [inline v0/destinations/am/transform.js:534]: name - may be `prepareTraitsConfig` ?

File: v0/destinations/am/transform.js [hunk 2]
```diff
@@ -516,17 +516,75 @@ function trackRevenueEvent(message, destination) {
...
+) {
+  let updatedTrait;
+  configPropertyTrait.forEach(traitCount => {
+    const property = traitCount.traits;
+    updatedTrait = updateTraitsObject(property, traitObject, objectToInclude);
+  });
+  return updatedTrait;
+}
+
+function traitsHandler(message, destination) {
```
Comment (prabrishac) [inline v0/destinations/am/transform.js:547]: name - may be `handleTraits` ?

File: v0/destinations/am/transform.js [hunk 3]
```diff
@@ -516,17 +516,75 @@ function trackRevenueEvent(message, destination) {
...
     destination.Config.trackProductsOnce === false
   ) {
-    const productPurchasedEvents = getProductPurchasedEvents(
-      message,
-      destination
-    );
-    if (productPurchasedEvents.length > 0) {
-      sendEvents = [...sendEvents, ...productPurchasedEvents];
-    }
+    sendEvents = getProductPurchasedEvents(message, destination, sendEvents);
```
Comment (prabrishac) [inline v0/destinations/am/transform.js:519]: remove the revenue related change or rebase it.

File: v0/destinations/am/transform.js [hunk 4]
```diff
@@ -516,17 +516,75 @@ function trackRevenueEvent(message, destination) {
...
+      traitObject.$add,
+      traitObject
+    );
+  }
+  if (destination.Config.traitsToSetOnce) {
+    traitObject.$setOnce = new Object();
+    updatedTrait = passingTraitConfigs(
+      destination.Config.traitsToSetOnce,
+      traitObject.$setOnce,
+      traitObject
```
Comment (prabrishac) [inline v0/destinations/am/transform.js:565]: can we pass only `traitObject` and the other parameter as string like, `$setOnce` or `$add` and prepare the same in called method?

**Review by prabrishac — CHANGES_REQUESTED** [review]

File: v0/destinations/am/transform.js [hunk 5]
```diff
@@ -541,6 +603,19 @@ function process(event) {
     } else {
       toSendEvents.push(message);
     }
+  } else if (messageType === EventType.IDENTIFY) {
```
Comment (prabrishac) [inline v0/destinations/am/transform.js:606]: we can move this code in `responseBuilderSimple` method.

File: v0/destinations/am/transform.js [hunk 6]
```diff
@@ -527,6 +524,71 @@ function trackRevenueEvent(message, destination) {
...
+  let updatedTrait;
+  traitsObject[objectToInclude] = new Object();
+  configTraitsPropertyKey.forEach(traitCount => {
+    const property = traitCount.traits;
+    updatedTrait = updateTraitsObject(property, traitsObject, objectToInclude);
+  });
+  return updatedTrait;
+}
+
+function handlingTraits(message, destination) {
```
Comment (prabrishac) [inline v0/destinations/am/transform.js:551]: rather, `handleTraits` ?

File: v0/destinations/am/transform.js [hunk 7]
```diff
@@ -527,6 +524,71 @@ function trackRevenueEvent(message, destination) {
...
+  return traitsObject;
+}
+
+function prepareTraitsConfig(
+  configTraitsPropertyKey,
+  objectToInclude,
+  traitsObject
+) {
+  let updatedTrait;
+  traitsObject[objectToInclude] = new Object();
```
Comment (prabrishac) [inline v0/destinations/am/transform.js:543]: please remove the lint error

File: v0/destinations/am/transform.js [hunk 8]
```diff
@@ -527,6 +524,71 @@ function trackRevenueEvent(message, destination) {
   return sendEvents;
 }
 
+function updateTraitsObject(property, traitsObject, newProperty) {
+  let propertyToUpdate;
+  if (traitsObject.hasOwnProperty(property)) {
```
Comment (prabrishac) [inline v0/destinations/am/transform.js:529]: please remove the lint error.

File: v0/destinations/am/transform.js [hunk 9]
```diff
@@ -527,6 +524,71 @@ function trackRevenueEvent(message, destination) {
...
+  if (destination.Config.traitsToIncrement) {
+    const objectToInclude = "$add";
+    updatedTrait = prepareTraitsConfig(
+      destination.Config.traitsToIncrement,
+      objectToInclude,
+      traitsObject
+    );
+  }
+  if (destination.Config.traitsToSetOnce) {
+    const objectToInclude = "$setOnce";
```
Comment (prabrishac) [inline v0/destinations/am/transform.js:565]: `objectToInclude` -> `actionKey`

Comment (prabrishac) [reply v0/destinations/am/transform.js:565]: replace the same in other places.

**Review by prabrishac — CHANGES_REQUESTED** [review]

**Review by prabrishac — APPROVED** [review]
Looks good to me 👍

**Comment by sayan-mitra** [comment]
since the changes are small, let's squash the commits

**Pull request merged**